### ISSUE TITLE
Improve Command debug printing on Windows

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#114385]
Debug-printing a `Command` on Unix system nicely prints not only the command itself but also the working directory and environment variables, if those were adjusted for this `Command`.

On Windows, none of this happens:

https://github.com/rust-lang/rust/blob/866710c552e5a9b9ce6d0f8143ebe8d69f7dd451/library/std/src/sys/windows/process.rs#L357-L369

We can probably do better here. The hardest part, I assume, is coming up with a suitable syntax -- `VAR=val program` unfortunately does not seem to be valid on Windows. Cargo [uses](https://github.com/rust-lang/cargo/blob/16d0495d297e1dc9b6a7d7827c0ea561d2485a02/crates/cargo-util/src/process_builder.rs#L55) `set VAR=val&& program` -- no idea of that would be a reasonable choice? Also, how should the working directory be printed? I know basically nothing about the Windows console so I hope others can chime in here. ;)

**Comment by luqmana** [comment]
`set VAR=val&& program` would work but the env variable would persist beyond just that invocation so it's not quite the same as `VAR=val program`. To remove a variable, `set VAR=`.

For working directory, the same `cd {cwd} &&` construct would work. That will also persist though that's true on unix as well, so maybe that's fine? (`pushd/popd` also works with `cmd.exe`).

If the goal is something you can just copy paste mostly, then might be worth wrapping with an explicit `cmd.exe` call so that it would also work in PowerShell:
```cmd
cmd.exe /c "cd {cwd} && set FOO=bar && set BAZ= && program"
```

**Comment by RalfJung** [comment]
> . To remove a variable, set VAR=.

If that's removing a variable, then how do I set it to the empty value? `cmd.env("VAR", "")` and `cmd.remove_env("VAR")` should be distinguishable.

> might be worth wrapping with an explicit cmd.exe

That looks like it will be a total disaster in terms of escaping. Arguments will already be printed with quotes around them (I think).

**Comment by nico-abram** [comment]
> `cmd.env("VAR", "")` and `cmd.remove_env("VAR")` should be distinguishable.

On windows, the following program:

```rs
fn main() {

	use std::process::Command;

	let output = Command::new("cmd")
            .env("test", "test")
            .args(["/C", "echo %test%"])
            .spawn()
            .expect("failed to execute process");

	let output = Command::new("cmd")
            .env("test", "")
            .args(["/C", "echo %test%"])
            .spawn()
            .expect("failed to execute process");
	println!("");
}
```

prints out (On `rustc 1.65.0 (897e37553 2022-11-02)`):
```rs
test
%test%
```

So I don't think those are distinguishable as implemented by std today

**Comment by RalfJung** [comment]
So does `std::env::var_os("NAME").is_some_and(|val| val.is_empty())` never return true on Windows?

**Comment by luqmana** [comment]
> So does `std::env::var_os("NAME").is_some_and(|val| val.is_empty())` never return true on Windows?

No, it can output `true`. AFAIK the underlying win32 APIs do allow an empty variable. But not sure if there's a way to do it in `cmd.exe` with `set`.

[Powershell just seems to straight up not allow it:](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.3#long-description)
>Environment variables, unlike other types of variables in PowerShell, are always stored as a string and can't be empty.
>
>Because an environment variable can't be an empty string, setting one to $null or an empty string removes it.

```rust
fn main() {
    std::env::set_var("NAME", "");
    println!("set and empty: {}", std::env::var_os("NAME").is_some_and(|val| val.is_empty()));
    // try var expansion
    std::process::Command::new("cmd").args(["/c", "echo", "%NAME%"]).spawn().unwrap();
    // print all env vars
    std::process::Command::new("cmd").args(["/c", "set"]).spawn().unwrap();
}
```
That outputs:
```
set and empty: true
%NAME%
[...snip other env vars...]
NAME=
```

**Comment by RalfJung** [comment]
But that means for Command, it does matter if the variable is empty vs absent, since the called program might be able to distinguish these cases. Just the shell cannot show the difference. (FWIW, on bash it is also hard to distinguish empty from absent. But it is possible.) We cannot have actually working syntax for debug-printing the empty case (or the absent case?), we have to come up with a different pseudo syntax.

When you call "set VAR= && prog", and " prog" is a Rust program that debug-prints env::var_os("VAR"), does it print None or Some("")? As in, does this remove the var or empty it?

**Comment by RalfJung** [comment]
Another action that would ideally be represented somehow is clearing the environment entirely. There's probably no command for that, but we can still print `clearenv && ...` to clarify that this is what happens. That seems better than not representing this information at all.

**Comment by ChrisDenton** [comment]
Here's all the info we may want in the debug output, depending on how `Command` is used:

* application name
* command line (where `argv[0]` may differ from the application name)
* security descriptor for the process
* security descriptor for the main thread
* [creation flags](https://learn.microsoft.com/en-us/windows/win32/procthread/process-creation-flags)
* environment variables (plus if the environment has been cleared)
* current directory
* [STARTUPINFOW](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/ns-processthreadsapi-startupinfow)
* [ProcThreadAttributes](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-updateprocthreadattribute)
